bootstrap/updater.go

```diff
@@ -0,0 +1,407 @@
+// Package bootstrap checks for the bootstrap/fallback data update from the
+// spacemesh administrator (a centralized entity controlled by the spacemesh
+// team). This is intended as a short-term solution at the beginning of the
+// network deployment to facilitate recovering from network failures and
+// should be removed once the network is stable.
+//
+// The updater periodically checks for the latest update from a URL provided
+// by the spacemesh administrator, verifies the data, persists on disk and
+// notifies subscribers of a new update.
+//
+// Subscribers register by calling `Subscribe()` to receive a channel for
+// the latest update.
+package bootstrap
+
+import (
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/santhosh-tekuri/jsonschema/v5"
+	"github.com/spf13/afero"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+)
+
+const (
+	DefaultURL = "http://localhost:3000/bootstrap"
+	DirName    = "bootstrap"
+
+	httpTimeout   = 5 * time.Second
+	notifyTimeout = time.Second
+	schemaFile    = "schema.json"
+	format        = "2006-01-02T15-04-05"
+)
+
+var (
+	ErrEpochOutOfOrder = errors.New("epoch out of order")
+	ErrWrongVersion    = errors.New("wrong schema version")
+	ErrInvalidBeacon   = errors.New("invalid beacon")
+)
+
+type Config struct {
+	URL     string `mapstructure:"bootstrap-uri"`
+	Version string `mapstructure:"bootstrap-version"`
+
+	DataDir   string
+	Interval  time.Duration
+	NumToKeep int
+}
+
+func DefaultConfig() Config {
+	return Config{
+		URL:       DefaultURL,
+		Version:   "https://spacemesh.io/bootstrap.schema.json.1.0",
+		DataDir:   os.TempDir(),
+		Interval:  30 * time.Second,
+		NumToKeep: 10,
+	}
+}
+
+type Updater struct {
+	cfg    Config
+	logger log.Log
+	fs     afero.Fs
+	client *http.Client
+	once   sync.Once
+	eg     errgroup.Group
+
+	mu          sync.Mutex
+	subscribers []chan *VerifiedUpdate
+	latest      *VerifiedUpdate
+}
+
+type Opt func(*Updater)
+
+func WithConfig(cfg Config) Opt {
+	return func(u *Updater) {
+		u.cfg = cfg
+	}
+}
+
+func WithLogger(logger log.Log) Opt {
+	return func(u *Updater) {
+		u.logger = logger
+	}
+}
+
+func WithFilesystem(fs afero.Fs) Opt {
+	return func(u *Updater) {
+		u.fs = fs
+	}
+}
+
+func WithHttpClient(c *http.Client) Opt {
+	return func(u *Updater) {
+		u.client = c
+	}
+}
+
+func New(opts ...Opt) *Updater {
+	u := &Updater{
+		cfg:    DefaultConfig(),
+		logger: log.NewNop(),
+		fs:     afero.NewOsFs(),
+		client: &http.Client{},
+	}
+	for _, opt := range opts {
+		opt(u)
+	}
+	return u
+}
+
+func (u *Updater) Subscribe() chan *VerifiedUpdate {
+	u.mu.Lock()
+	defer u.mu.Unlock()
+	ch := make(chan *VerifiedUpdate, 10)
+	u.subscribers = append(u.subscribers, ch)
```

Comment:
Do we need to push the latest update to the newly subscribed channel?

Reply:
this is a design decision. there is no need for now. but i'll keep that in mind in the upcoming iterations.

Reply:
nope. if in my follow up PRs it looks problematic to you, then i can change then.